### PR TITLE
Fix Literal values

### PR DESCRIPTION
**Pull request by KevinRansom**
This PR:  https://github.com/Microsoft/visualfsharp/pull/4272

introduced a bug in code generation for literal fields.  Instead of writing them is a literal it wrote them as a value.

This issue is not in VS 2017.6 it only appears in VS2017.7 preview 1 internal builds.

The fix is to ensure that the FieldDef knows about the literal.

/cc @dsyme, @AviAvni, @jaredpar

### CODE REVIEW

**Comment by AviAvni** [comment]
why ILComparer didn't catch this?

**Comment by KevinRansom** [comment]
@AviAvni --- we were surprised none of the tests caught it but for sure they didn't catch it.  Perhaps a regression test would be good.

**Comment by AviAvni** [comment]
@KevinRansom so is there sample code that failed?
I'll send pr with the test and fixed WithLiteral to get litaral value instead of bool as @forki suggested

**Comment by dsyme** [comment]
@aviavni yes please: an ILComparer test for this case and adjusted WithLiteral.

**Comment by KevinRansom** [comment]
````
module Foo =

    [<Literal>]
    let x = 7
````

**Comment by forki** [comment]
@dsyme is it already rel in FCS? If so we need to notify fable and fez

**Comment by KevinRansom** [comment]
@forki yes, the bug exists in FCS:   

https://github.com/fsharp/FSharp.Compiler.Service/blob/master/src/fsharp/IlxGen.fs#L4745

**Comment by forki** [comment]
/cc @alfonsogcnunez @kjnilsson

Question is if it's already on nuget

**Comment by KevinRansom** [comment]
@alfonsogarciacaro there is a bug in field generation and that has made it into the FCS.  We have a fix that works, and a proposal for a better one "less weird looking code :-)"

You might want to make sure and wait too update to the latest FCS if you depend on [<Literal>] making a difference to your code gen strategy for code that looks like:

````

module Stuff =
   [<Literal>] 
   let x = 7

**Comment by forki** [comment]
@kevinransom that was not meant as criticism of your fix :-) just as a
suggestion to merge both bits into one set function so that nobody can ever
do it wrong again

Am 11.03.2018 19:30 schrieb "Kevin Ransom (msft)" <notifications@github.com

**Comment by KevinRansom** [comment]
Lol … you were right it does look weird my friend, and I like the suggestion of course I pushed up the PR as soon as I had something that worked,  given that 17.7 preview 1 has already shipped and I wanted to see if we can get an update in the channel rapidly.

**Comment by KevinRansom** [comment]
@forki --

the last update to nuget was 9 days ago.
And the change was merged into FCS 12 days ago:  https://github.com/fsharp/FSharp.Compiler.Service/commit/a541ef7d92199ab9bcc1db685972d05e709b7348

I think the chances are good that the latest FCS has the problem, but it's only serious if you depend on [<Literal>] changing the code gen strategy.

**Comment by forki** [comment]
Well that's why I want to let fez and fable know...

Am 11.03.2018 19:45 schrieb "Kevin Ransom (msft)" <notifications@github.com

**Comment by dsyme** [comment]
@forki I don't think this affects either fez or fable.   Only tools writing .NET binaries are affected.

**Comment by forki** [comment]
ah yes it's in IL*. Didn't see that. So AST is fine. Good!

File: src/absil/il.fs [hunk 1]
```diff
@@ -1558,7 +1558,7 @@ type ILFieldDef =
     member x.WithStatic(condition) = { x with Attributes = x.Attributes |> conditionalAdd condition FieldAttributes.Static }
     member x.WithSpecialName(condition) = { x with Attributes = x.Attributes |> conditionalAdd condition (FieldAttributes.SpecialName ||| FieldAttributes.RTSpecialName) }
     member x.WithNotSerialized(condition) = { x with Attributes = x.Attributes |> conditionalAdd condition FieldAttributes.NotSerialized }
-    member x.WithLiteral(condition) = { x with Attributes = x.Attributes |> conditionalAdd condition FieldAttributes.Literal }
+    member x.WithLiteralDefaultValue(literal) = { x with Attributes = x.Attributes |> conditionalAdd (isSome literal) FieldAttributes.Literal ||| FieldAttributes.HasDefault}
```
Comment (forki) [inline src/absil/il.fs]: @KevinRansom  shouldn't this also set the LiteralValue?

Comment (KevinRansom) [reply src/absil/il.fs]: @forki, it should, but I got excited about incorporating DefaultValue as well, anyway it doesn't even compile. So clearly more work is required.

Comment (forki) [reply src/absil/il.fs]: ;-)

**Comment by KevinRansom** [comment]
@AviAvni   I did a bit of a refactor, can you check it out and let me know if I did it right?

Thanks

Kevin

File: src/fsharp/IlxGen.fs [hunk 2]
```diff
@@ -4742,7 +4742,7 @@ and GenBindingAfterSequencePoint cenv cgbuf eenv sp (TBind(vspec,rhsExpr,_)) sta
             let ilFieldDef = mkILStaticField (fspec.Name, fty, None, None, access)
             let ilFieldDef =
                 match vref.LiteralValue with 
-                | Some konst -> { ilFieldDef.WithHasDefault(true) with LiteralValue = Some(GenFieldInit m konst) }
+                | Some konst -> ilFieldDef.WithLiteralDefaultValue( Some (GenFieldInit m konst) )
```
Comment (forki) [inline src/fsharp/IlxGen.fs]: why do we need to pass it an option? Is there a situation where passing None makes sense?

**Comment by AviAvni** [comment]
@KevinRansom look great

File: src/fsharp/IlxGen.fs (hunk 2 above)
Comment (AviAvni) [reply src/fsharp/IlxGen.fs]: no passing None but when using it is more comfort

**Comment by KevinRansom** [comment]
Okay .. closing this in favor of: 
https://github.com/Microsoft/visualfsharp/pull/4486
But I have to merge in these changes to it :-)

**Pull request closed without merging**